#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [41]:
import pandas as pd

data_path = './data/supermarket_sales.csv'
df = pd.read_csv(data_path)
df

Invoice ID Branch       City Customer type  Gender  \
0    750-67-8428      A     Yangon        Member  Female   
1    226-31-3081      C  Naypyitaw        Normal  Female   
2    631-41-3108      A     Yangon        Normal    Male   
3    123-19-1176      A     Yangon        Member    Male   
4    373-73-7910      A     Yangon        Normal    Male   
..           ...    ...        ...           ...     ...   
995  233-67-5758      C  Naypyitaw        Normal    Male   
996  303-96-2227      B   Mandalay        Normal  Female   
997  727-02-1313      A     Yangon        Member    Male   
998  347-56-2442      A     Yangon        Normal    Male   
999  849-09-3807      A     Yangon        Member  Female   

               Product line  Unit price  Quantity   Tax 5%      Total  \
0         Health and beauty       74.69         7  26.1415   548.9715   
1    Electronic accessories       15.28         5   3.8200    80.2200   
2        Home and lifestyle       46.33         7  16.2155   340.5255   
3         Health and beauty       58.22         8  23.2880   489.0480   
4         Sports and travel       86.31         7  30.2085   634.3785   
..                      ...         ...       ...      ...        ...   
995       Health and beauty       40.35         1   2.0175    42.3675   
996      Home and lifestyle       97.38        10  48.6900  1022.4900   
997      Food and beverages       31.84         1   1.5920    33.4320   
998      Home and lifestyle       65.82         1   3.2910    69.1110   
999     Fashion accessories       88.34         7  30.9190   649.2990   

          Date   Time      Payment    cogs  gross margin percentage  \
0     1/5/2019  13:08      Ewallet  522.83                 4.761905   
1     3/8/2019  10:29         Cash   76.40                 4.761905   
2     3/3/2019  13:23  Credit card  324.31                 4.761905   
3    1/27/2019  20:33      Ewallet  465.76                 4.761905   
4     2/8/2019  10:37      Ewallet  604.17                 4.761905   
..         ...    ...          ...     ...                      ...   
995  1/29/2019  13:46      Ewallet   40.35                 4.761905   
996   3/2/2019  17:16      Ewallet  973.80                 4.761905   
997   2/9/2019  13:22         Cash   31.84                 4.761905   
998  2/22/2019  15:33         Cash   65.82                 4.761905   
999  2/18/2019  13:28         Cash  618.38                 4.761905   

     gross income  Rating  
0         26.1415     9.1  
1          3.8200     9.6  
2         16.2155     7.4  
3         23.2880     8.4  
4         30.2085     5.3  
..            ...     ...  
995        2.0175     6.2  
996       48.6900     4.4  
997        1.5920     7.7  
998        3.2910     4.1  
999       30.9190     6.6  

[1000 rows x 17 columns]

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [7]:
df.groupby(by='Branch')['Unit price'].sum()

Branch
A    18625.49
B    18478.88
C    18567.76
Name: Unit price, dtype: float64

Найбільший прибуток спостерігається в філіалі А і дорівнює 18625.49

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [36]:
grouped = df.groupby(by=['City','Branch'])['gross income'].sum()
max_index = grouped.idxmax()
max_value = grouped[max_index]

print(grouped, '\n')

print(f"Місто в якому знаходиться філіал з найвищим прибутком це - {max_index[0]}, філіал: {max_index[1]}, прибуток: {max_value}")

City       Branch
Mandalay   B         5057.0320
Naypyitaw  C         5265.1765
Yangon     A         5057.1605
Name: gross income, dtype: float64 

Місто в якому знаходиться філіал з найвищим прибутком це - Naypyitaw, філіал: C, прибуток: 5265.1765


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [43]:
invoices_by_category = pd.pivot_table(df, index='Branch', columns='Product line', values='Invoice ID', aggfunc='count')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [75]:
invoices_by_category['Electronic accessories']

max_value = invoices_by_category['Electronic accessories'].max()
max_row_index = invoices_by_category["Electronic accessories"].idxmax()

print(f"Найбільше інвойсів у філіалу '{max_row_index}' у кількості  {max_value}")

Найбільше інвойсів у філіалу 'A' у кількості  60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


1. Гіпотези
   1.1 Які товари та в якому місті купуються найбільше?
   1.2 Хто та які товари купує найбільше (чоловіки / жінки)?
   1.3 Які групи товарів та в який час купуються? 

In [98]:
# 2.1 Які товари та в якому місті купуються найбільше?

#df.groupby(by=['Product line','City'])['Invoice ID'].count()

grouped = pd.pivot_table(df, index='City', columns='Product line', values='Invoice ID', aggfunc='count')
display(grouped)

max_values = grouped.max()
max_cities = grouped.idxmax()
result = pd.DataFrame({"Max Value": max_values, "City": max_cities})

print(result)

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
City                                                                            
Mandalay                          55                   62                  50   
Naypyitaw                         55                   65                  66   
Yangon                            60                   51                  58   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
City                                                                    
Mandalay                     53                  50                 62  
Naypyitaw                    52                  45                 45  
Yangon                       47                  65                 59

                        Max Value       City
Product line                                
Electronic accessories         60     Yangon
Fashion accessories            65  Naypyitaw
Food and beverages             66  Naypyitaw
Health and beauty              53   Mandalay
Home and lifestyle             65     Yangon
Sports and travel              62   Mandalay


За результатами отриманих даних можна зробити висновки які групи товарів та в якому місті купуються найкраще і відповідно до них будувати маркетингову стратегію.

За результатами видно, що 
- в місті Naypyitaw краще продаються Fashion accessories та Food and beverages
- в місті Yangon краще продаються Electronic accessories та Home and lifestyle
- в місті Mandalay краще продаються Health and beauty та Sports and travel

In [100]:
# 2.2 Хто та які товари купує найбільше (чоловіки / жінки)?

grouped = pd.pivot_table(df, index='Gender', columns='Product line', values='Invoice ID', aggfunc='count')
display(grouped)

max_values = grouped.max()
max_cities = grouped.idxmax()
result = pd.DataFrame({"Max Value": max_values, "Gender": max_cities})

print(result)

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Gender                                                                          
Female                            84                   96                  90   
Male                              86                   82                  84   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Gender                                                                  
Female                       64                  79                 88  
Male                         88                  81                 78

                        Max Value  Gender
Product line                             
Electronic accessories         86    Male
Fashion accessories            96  Female
Food and beverages             90  Female
Health and beauty              88    Male
Home and lifestyle             81    Male
Sports and travel              88  Female


За результатами отриманих даних видно що 
- жінки більше купують товари Fashion accessories, Food and beverages та Sports and travel
- чоловіків більше цікавлять тьовари з груп Electronic accessories, Health and beauty та Home and lifestyle.

За допомогою цих даних можна розробляти рекламу розуміючи цільову категорію.

In [102]:
# 2.3 Які групи товарів та в який час купуються? 

def get_hour(time_value):
    hour = time_value.split(':')[0]
    return int(hour)

df["Hour"] = df.Time.apply(lambda x: get_hour(x))
df["Hour"].value_counts()

grouped = pd.pivot_table(df, index='Hour', columns='Product line', values='Invoice ID', aggfunc='count')
display(grouped)

max_values = grouped.max()
max_cities = grouped.idxmax()
result = pd.DataFrame({"Max Value": max_values, "Hour": max_cities})

print(result)

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Hour                                                                            
10                                17                   14                  20   
11                                18                   14                  11   
12                                16                   19                  10   
13                                13                   19                  15   
14                                14                   15                  12   
15                                15                   18                  21   
16                                11                   12                  16   
17                                14                   13                   6   
18                                20                   19                  17   
19                                16                   24                  26   
20                                16                   11                  20   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Hour                                                                    
10                           17                  18                 15  
11                           10                  21                 16  
12                           14                  14                 16  
13                           17                  20                 19  
14                           19                   9                 14  
15                           12                  19                 17  
16                           10                   8                 20  
17                           11                  17                 13  
18                           14                  11                 12  
19                           14                  13                 20  
20                           14                  10                  4

                        Max Value  Hour
Product line                           
Electronic accessories         20    18
Fashion accessories            24    19
Food and beverages             26    19
Health and beauty              19    14
Home and lifestyle             21    11
Sports and travel              20    16


На основі отриманих результатів можна бачити в які години краще продаються ті чи інші товари та відповідно робити різні акції типу "щасливих годин" для підвищення попиту а відповідно і продажів.